### ISSUE TITLE
Use of a type parameter pack where one or more type parameters are required results in a compiler crash.

### ISSUE THREAD

**Issue by jamesmmadeira** [swiftlang/swift#68159]
**Description**
Whenever I use a type parameter pack in a place that requires more than one type parameter, the compiler crashes.

<details><summary>Crash backtrace when building in Xcode</summary>

```
Please submit a bug report (https://swift.org/contributing/#reporting-bugs) and include the crash backtrace.
Stack dump:
0.	Program arguments: /Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/swift-frontend -frontend -emit-module -experimental-skip-non-inlinable-function-bodies-without-types "/Users/jamesmadeira/Library/Mobile Documents/com~apple~CloudDocs/Developer/CrashTest/CrashTest/main.swift" -target arm64-apple-macos14.0 -Xllvm -aarch64-use-tbi -enable-objc-interop -stack-check -sdk /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk -I /Users/jamesmadeira/Library/Developer/Xcode/DerivedData/CrashTest-arpsknbrjiutkreycvszovrpqgfx/Build/Products/Debug -F /Users/jamesmadeira/Library/Developer/Xcode/DerivedData/CrashTest-arpsknbrjiutkreycvszovrpqgfx/Build/Products/Debug -no-color-diagnostics -enable-testing -g -module-cache-path /Users/jamesmadeira/Library/Developer/Xcode/DerivedData/ModuleCache.noindex -swift-version 5 -enforce-exclusivity=checked -Onone -D DEBUG -serialize-debugging-options -const-gather-protocols-file /Users/jamesmadeira/Library/Developer/Xcode/DerivedData/CrashTest-arpsknbrjiutkreycvszovrpqgfx/Build/Intermediates.noindex/CrashTest.build/Debug/CrashTest.build/Objects-normal/arm64/CrashTest_const_extract_protocols.json -enable-bare-slash-regex -empty-abi-descriptor -validate-clang-modules-once -clang-build-session-file /Users/jamesmadeira/Library/Developer/Xcode/DerivedData/ModuleCache.noindex/Session.modulevalidation -Xcc -working-directory -Xcc "/Users/jamesmadeira/Library/Mobile Documents/com~apple~CloudDocs/Developer/CrashTest" -resource-dir /Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift -enable-anonymous-context-mangled-names -Xcc -ivfsstatcache -Xcc /Users/jamesmadeira/Library/Developer/Xcode/DerivedData/SDKStatCaches.noindex/macosx14.0-23A5326c-f7e79a2b09dbe2c8a4ba251adb57f722.sdkstatcache -Xcc -I/Users/jamesmadeira/Library/Developer/Xcode/DerivedData/CrashTest-arpsknbrjiutkreycvszovrpqgfx/Build/Intermediates.noindex/CrashTest.build/Debug/CrashTest.build/swift-overrides.hmap -Xcc -iquote -Xcc /Users/jamesmadeira/Library/Developer/Xcode/DerivedData/CrashTest-arpsknbrjiutkreycvszovrpqgfx/Build/Intermediates.noindex/CrashTest.build/Debug/CrashTest.build/CrashTest-generated-files.hmap -Xcc -I/Users/jamesmadeira/Library/Developer/Xcode/DerivedData/CrashTest-arpsknbrjiutkreycvszovrpqgfx/Build/Intermediates.noindex/CrashTest.build/Debug/CrashTest.build/CrashTest-own-target-headers.hmap -Xcc -I/Users/jamesmadeira/Library/Developer/Xcode/DerivedData/CrashTest-arpsknbrjiutkreycvszovrpqgfx/Build/Intermediates.noindex/CrashTest.build/Debug/CrashTest.build/CrashTest-all-target-headers.hmap -Xcc -iquote -Xcc /Users/jamesmadeira/Library/Developer/Xcode/DerivedData/CrashTest-arpsknbrjiutkreycvszovrpqgfx/Build/Intermediates.noindex/CrashTest.build/Debug/CrashTest.build/CrashTest-project-headers.hmap -Xcc -I/Users/jamesmadeira/Library/Developer/Xcode/DerivedData/CrashTest-arpsknbrjiutkreycvszovrpqgfx/Build/Products/Debug/include -Xcc -I/Users/jamesmadeira/Library/Developer/Xcode/DerivedData/CrashTest-arpsknbrjiutkreycvszovrpqgfx/Build/Intermediates.noindex/CrashTest.build/Debug/CrashTest.build/DerivedSources-normal/arm64 -Xcc -I/Users/jamesmadeira/Library/Developer/Xcode/DerivedData/CrashTest-arpsknbrjiutkreycvszovrpqgfx/Build/Intermediates.noindex/CrashTest.build/Debug/CrashTest.build/DerivedSources/arm64 -Xcc -I/Users/jamesmadeira/Library/Developer/Xcode/DerivedData/CrashTest-arpsknbrjiutkreycvszovrpqgfx/Build/Intermediates.noindex/CrashTest.build/Debug/CrashTest.build/DerivedSources -Xcc -DDEBUG=1 -module-name CrashTest -disable-clang-spi -target-sdk-version 14.0 -target-sdk-name macosx14.0 -external-plugin-path /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/lib/swift/host/plugins#/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/bin/swift-plugin-server -external-plugin-path /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/local/lib/swift/host/plugins#/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/bin/swift-plugin-server -external-plugin-path /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/usr/lib/swift/host/plugins#/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/usr/bin/swift-plugin-server -external-plugin-path /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/usr/local/lib/swift/host/plugins#/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/usr/bin/swift-plugin-server -plugin-path /Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/host/plugins -plugin-path /Applications/Xcode-beta.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/local/lib/swift/host/plugins -emit-module-doc-path /Users/jamesmadeira/Library/Developer/Xcode/DerivedData/CrashTest-arpsknbrjiutkreycvszovrpqgfx/Build/Intermediates.noindex/CrashTest.build/Debug/CrashTest.build/Objects-normal/arm64/CrashTest.swiftdoc -emit-module-source-info-path /Users/jamesmadeira/Library/Developer/Xcode/DerivedData/CrashTest-arpsknbrjiutkreycvszovrpqgfx/Build/Intermediates.noindex/CrashTest.build/Debug/CrashTest.build/Objects-normal/arm64/CrashTest.swiftsourceinfo -emit-objc-header-path /Users/jamesmadeira/Library/Developer/Xcode/DerivedData/CrashTest-arpsknbrjiutkreycvszovrpqgfx/Build/Intermediates.noindex/CrashTest.build/Debug/CrashTest.build/Objects-normal/arm64/CrashTest-Swift.h -serialize-diagnostics-path /Users/jamesmadeira/Library/Developer/Xcode/DerivedData/CrashTest-arpsknbrjiutkreycvszovrpqgfx/Build/Intermediates.noindex/CrashTest.build/Debug/CrashTest.build/Objects-normal/arm64/CrashTest-master-emit-module.dia -emit-dependencies-path /Users/jamesmadeira/Library/Developer/Xcode/DerivedData/CrashTest-arpsknbrjiutkreycvszovrpqgfx/Build/Intermediates.noindex/CrashTest.build/Debug/CrashTest.build/Objects-normal/arm64/CrashTest-master-emit-module.d -o /Users/jamesmadeira/Library/Developer/Xcode/DerivedData/CrashTest-arpsknbrjiutkreycvszovrpqgfx/Build/Intermediates.noindex/CrashTest.build/Debug/CrashTest.build/Objects-normal/arm64/CrashTest.swiftmodule -emit-abi-descriptor-path /Users/jamesmadeira/Library/Developer/Xcode/DerivedData/CrashTest-arpsknbrjiutkreycvszovrpqgfx/Build/Intermediates.noindex/CrashTest.build/Debug/CrashTest.build/Objects-normal/arm64/CrashTest.abi.json
1.	Apple Swift version 5.9 (swiftlang-5.9.0.128.106 clang-1500.0.40.1)
2.	Compiling with the current language version
3.	While evaluating request TypeCheckSourceFileRequest(source_file "/Users/jamesmadeira/Library/Mobile Documents/com~apple~CloudDocs/Developer/CrashTest/CrashTest/main.swift")
4.	While type-checking 'f(_:)' (at /Users/jamesmadeira/Library/Mobile Documents/com~apple~CloudDocs/Developer/CrashTest/CrashTest/main.swift:10:1)
5.	While evaluating request InterfaceTypeRequest(CrashTest.(file).f@/Users/jamesmadeira/Library/Mobile Documents/com~apple~CloudDocs/Developer/CrashTest/CrashTest/main.swift:10:6)
6.	While evaluating request GenericSignatureRequest(CrashTest.(file).f@/Users/jamesmadeira/Library/Mobile Documents/com~apple~CloudDocs/Developer/CrashTest/CrashTest/main.swift:10:6)
7.	While evaluating request ResolveTypeRequest(while resolving type , S<repeat each T>)
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  swift-frontend           0x00000001033377b4 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 56
1  swift-frontend           0x0000000105eb5f0c llvm::sys::RunSignalHandlers() + 112
2  swift-frontend           0x0000000101bf7720 SignalHandler(int) + 352
3  libsystem_platform.dylib 0x00000001823c6a24 _sigtramp + 56
4  swift-frontend           0x000000010326ccbc applyGenericArguments(swift::Type, swift::TypeResolution, swift::SILTypeResolutionContext*, swift::IdentTypeRepr*) + 3340
5  swift-frontend           0x0000000105cd6380 resolveTypeDecl(swift::TypeDecl*, swift::DeclContext*, swift::TypeResolution, swift::SILTypeResolutionContext*, swift::IdentTypeRepr*) + 196
6  swift-frontend           0x000000010325d100 (anonymous namespace)::TypeResolver::resolveType(swift::TypeRepr*, swift::TypeResolutionOptions) + 2968
7  swift-frontend           0x0000000105cbfa04 evaluateTypeResolution(swift::TypeResolution const*, swift::TypeRepr*, swift::SILTypeResolutionContext*) (.llvm.5495460461374514410) + 144
8  swift-frontend           0x0000000103271168 llvm::Expected<swift::ResolveTypeRequest::OutputType> swift::Evaluator::getResultUncached<swift::ResolveTypeRequest>(swift::ResolveTypeRequest const&) + 476
9  swift-frontend           0x000000010325aec0 swift::TypeResolution::resolveType(swift::TypeRepr*, swift::SILTypeResolutionContext*) const + 156
10 swift-frontend           0x00000001058881f0 swift::GenericSignatureRequest::evaluate(swift::Evaluator&, swift::GenericContext*) const + 3600
11 swift-frontend           0x000000010387fbb4 swift::GenericContext::getGenericSignature() const + 1620
12 swift-frontend           0x0000000102f49d4c swift::InterfaceTypeRequest::evaluate(swift::Evaluator&, swift::ValueDecl*) const + 284
13 swift-frontend           0x00000001038a8728 swift::InterfaceTypeRequest::OutputType swift::evaluateOrDefault<swift::InterfaceTypeRequest>(swift::Evaluator&, swift::InterfaceTypeRequest, swift::InterfaceTypeRequest::OutputType) + 1028
14 swift-frontend           0x0000000102ffe7c8 (anonymous namespace)::DeclChecker::visitFuncDecl(swift::FuncDecl*) + 156
15 swift-frontend           0x00000001057cc604 (anonymous namespace)::DeclChecker::visit(swift::Decl*) (.llvm.14083370151234498882) + 1112
16 swift-frontend           0x0000000102ff5e94 swift::TypeChecker::typeCheckDecl(swift::Decl*, bool) + 220
17 swift-frontend           0x0000000103276040 swift::TypeCheckSourceFileRequest::evaluate(swift::Evaluator&, swift::SourceFile*) const + 432
18 swift-frontend           0x000000010327e2ec llvm::Expected<swift::TypeCheckSourceFileRequest::OutputType> swift::Evaluator::getResultUncached<swift::TypeCheckSourceFileRequest>(swift::TypeCheckSourceFileRequest const&) + 664
19 swift-frontend           0x0000000103275cc8 swift::TypeCheckSourceFileRequest::OutputType swift::evaluateOrDefault<swift::TypeCheckSourceFileRequest>(swift::Evaluator&, swift::TypeCheckSourceFileRequest, swift::TypeCheckSourceFileRequest::OutputType) + 228
20 swift-frontend           0x0000000101ee33c0 swift::CompilerInstance::performSema() + 296
21 swift-frontend           0x0000000105e6d3f8 performCompile(swift::CompilerInstance&, int&, swift::FrontendObserver*) + 464
22 swift-frontend           0x0000000105e71454 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 4568
23 swift-frontend           0x00000001057b74c0 swift::mainEntry(int, char const**) + 4116
24 dyld                     0x000000018203ff28 start + 2236
Command SwiftEmitModule failed with a nonzero exit code
```
</details> 

<details><summary>Crash backtrace when building in SwiftFiddle</summary>

```
Swift version 5.9-dev (LLVM 46ac717ed0ec21c, Swift 05fe36c7b352712)              21:23:51 10s
Target: x86_64-unknown-linux-gnu
swift-frontend: /home/build-user/swift/lib/Sema/TypeCheckType.cpp:918: swift::Type applyGenericArguments(swift::Type, swift::TypeResolution, swift::SILTypeResolutionContext *, swift::IdentTypeRepr *): Assertion `found != matcher.pairs.end()' failed.
Stack dump:
0.      Program arguments: /usr/bin/swift-frontend -frontend -interpret /main.swift -disable-objc-interop -I swiftfiddle.com/_Packages/.build/release -color-diagnostics -new-driver-path /usr/bin/swift-driver -empty-abi-descriptor -resource-dir /usr/lib/swift -module-name main -plugin-path /usr/lib/swift/host/plugins -plugin-path /usr/local/lib/swift/host/plugins -l_Packages
1.      Swift version 5.9-dev (LLVM 46ac717ed0ec21c, Swift 05fe36c7b352712)
2.      Compiling with the current language version
3.      While evaluating request TypeCheckSourceFileRequest(source_file "/main.swift")
4.      While type-checking 'f(_:)' (at /main.swift:7:1)
5.      While evaluating request InterfaceTypeRequest(main.(file).f@/main.swift:7:6)
6.      While evaluating request GenericSignatureRequest(main.(file).f@/main.swift:7:6)
7.      While evaluating request ResolveTypeRequest(while resolving type , S<repeat each T>)
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
/usr/bin/swift-frontend(+0x6fdbec3)[0x5638f81c0ec3]
/usr/bin/swift-frontend(+0x6fd9bbe)[0x5638f81bebbe]
/usr/bin/swift-frontend(+0x6fdc23f)[0x5638f81c123f]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x14420)[0x7fc8b1f92420]
/lib/x86_64-linux-gnu/libc.so.6(gsignal+0xcb)[0x7fc8b13c600b]
/lib/x86_64-linux-gnu/libc.so.6(abort+0x12b)[0x7fc8b13a5859]
/lib/x86_64-linux-gnu/libc.so.6(+0x22729)[0x7fc8b13a5729]
/lib/x86_64-linux-gnu/libc.so.6(+0x33fd6)[0x7fc8b13b6fd6]
/usr/bin/swift-frontend(+0x22b6d2e)[0x5638f349bd2e]
/usr/bin/swift-frontend(+0x22b3d41)[0x5638f3498d41]
/usr/bin/swift-frontend(+0x22abd70)[0x5638f3490d70]
/usr/bin/swift-frontend(+0x22a7a53)[0x5638f348ca53]
/usr/bin/swift-frontend(+0x22bb0a6)[0x5638f34a00a6]
/usr/bin/swift-frontend(+0x22a7bc8)[0x5638f348cbc8]
/usr/bin/swift-frontend(+0x22a79ea)[0x5638f348c9ea]
/usr/bin/swift-frontend(+0x21f9843)[0x5638f33de843]
/usr/bin/swift-frontend(+0x267ed09)[0x5638f3863d09]
/usr/bin/swift-frontend(+0x267eac9)[0x5638f3863ac9]
/usr/bin/swift-frontend(+0x2647d9b)[0x5638f382cd9b]
/usr/bin/swift-frontend(+0x218b5ea)[0x5638f33705ea]
/usr/bin/swift-frontend(+0x26a25fd)[0x5638f38875fd]
/usr/bin/swift-frontend(+0x26a23b9)[0x5638f38873b9]
/usr/bin/swift-frontend(+0x26523f8)[0x5638f38373f8]
/usr/bin/swift-frontend(+0x2644a44)[0x5638f3829a44]
/usr/bin/swift-frontend(+0x21cde03)[0x5638f33b2e03]
/usr/bin/swift-frontend(+0x21c1500)[0x5638f33a6500]
/usr/bin/swift-frontend(+0x21c1350)[0x5638f33a6350]
/usr/bin/swift-frontend(+0x22bc71a)[0x5638f34a171a]
/usr/bin/swift-frontend(+0x22bf1e4)[0x5638f34a41e4]
/usr/bin/swift-frontend(+0x22bef82)[0x5638f34a3f82]
/usr/bin/swift-frontend(+0x22bc584)[0x5638f34a1584]
/usr/bin/swift-frontend(+0x22bc564)[0x5638f34a1564]
/usr/bin/swift-frontend(+0x102bf49)[0x5638f2210f49]
/usr/bin/swift-frontend(+0x102178e)[0x5638f220678e]
/usr/bin/swift-frontend(+0x10216d7)[0x5638f22066d7]
/usr/bin/swift-frontend(+0xdad5de)[0x5638f1f925de]
/usr/bin/swift-frontend(+0xd9aa8b)[0x5638f1f7fa8b]
/usr/bin/swift-frontend(+0xd995fe)[0x5638f1f7e5fe]
/usr/bin/swift-frontend(+0xbf9143)[0x5638f1dde143]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3)[0x7fc8b13a7083]
/usr/bin/swift-frontend(+0xbf887e)[0x5638f1ddd87e]
Aborted (core dumped)
```
</details>

(Note that the SwiftFiddle backtrace includes the failed assertion, unlike the Xcode backtrace.)

**Steps to reproduce**
Try to compile the following program.
```swift
struct S<First, each Other> {}

func f<each T>(_: repeat each T) -> S<repeat each T> {
    fatalError()
}
```
[Here's the SwiftFiddle I used.](https://swiftfiddle.com/ygw3sntigfbtxnek4cb5qaww4m)

**Expected behavior**
Either:
1. The compilation succeeds and the compiler infers that `each T` must contain one or more types
2. The compilation fails and the compiler gives the programmer a helpful error instead of crashing

**Environment** (for Xcode backtrace)
- Swift compiler version info `swift-driver version: 1.87.1 Apple Swift version 5.9 (swiftlang-5.9.0.128.106 clang-1500.0.40.1)
Target: arm64-apple-macosx13.0`
- Xcode version info `Xcode 15.0
Build version 15A5229h` (Beta 7)
- Deployment target: macOS 14.0

**Comment by simanerush** [comment]
This does not crash anymore:
```
error: generic type 'S' specialized with mismatched type parameter pack
func f<each T>(_: repeat each T) -> S<repeat each T> {
                                    ^~~~~~~~~~~~~~~~
note: generic type 'S' declared here
struct S<First, each Other> {}
```

**Comment by AnthonyLatsis** [comment]
Fixed by #69116.
